File: 0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution 
{
    public bool IsValid(string s) 
    {
        if (s.Length % 2 != 0)
            return false;
        
        var bracketDictionary = new Dictionary<char, char>()
                              {
                                  { '(', ')' },
                                  { '[', ']' },
                                  { '{', '}' }
                              };
        
        var stack = new Stack<char>();
        
        foreach (var c in s)
        {
            if (bracketDictionary.ContainsKey(c))
                stack.Push(c);
            else
            {
                if (stack.Count == 0 || bracketDictionary[stack.Peek()] != c)
                    return false;
                
                stack.Pop();
            }
        }
        
        return stack.Count == 0;
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution 
{
    public:
        bool isValid(string s) 
        {
            unordered_map<char, char> bracket_pairs
			{
				{'(', ')'},
				{'[', ']'},
				{'{', '}'}
			};

			stack<char> stack;

			for(char c : s)
			{
				if(bracket_pairs.find(c) != bracket_pairs.end())
					stack.push(bracket_pairs[c]);
				
				else
				{
					if(stack.empty() || stack.top() != c)
						return false;

					stack.pop();
				}
			}

			return stack.empty();
        }
};